src/character_functions.cpp

```diff
@@ -595,63 +599,140 @@ bool try_wield_contents( Character &who, item &container, item *internal_item, b
     return true;
 }
 
+constexpr int trigger_rcs_cost_kcal = 10;
+
+
+auto uncanny_dodge_result( const Character &who ) -> UncannyDodgeResult
+{
+    const bool is_rcs_active = who.has_active_bionic( bio_uncanny_dodge_rcs );
+
+    if( is_rcs_active
+        && who.get_dodge() < 10 // pointless at 100% dodge
+        && who.get_stored_kcal() > trigger_rcs_cost_kcal ) {
+        const auto adjacent = pick_safe_adjacent_tile( who );
+        if( adjacent ) {
+            return *adjacent;
+        }
+    }
+
+    if( who.get_power_level() < bio_uncanny_dodge->power_trigger ) {
+        return UncannyDodgeStatus::NoEnergy;
+    }
+
+    if( x_in_y( 10 + who.get_dodge(), 20 ) ) {
+        return UncannyDodgeStatus::DodgedWithSkill;
+    }
+
+    return UncannyDodgeStatus::NoSpace;
+}
+
 bool try_uncanny_dodge( Character &who )
 {
-    const units::energy trigger_cost = bio_uncanny_dodge->power_trigger;
-    if( who.get_power_level() < trigger_cost || !who.has_active_bionic( bio_uncanny_dodge ) ) {
+    const bool is_active = who.has_active_bionic( bio_uncanny_dodge );
+
+    if( !is_active ) {
         return false;
     }
-    who.mod_power_level( -trigger_cost );
-    bool is_u = who.is_avatar();
-    bool seen = is_u || get_player_character().sees( who );
-    std::optional<tripoint> adjacent = pick_safe_adjacent_tile( who );
-    if( adjacent ) {
-        if( is_u ) {
-            add_msg( _( "Time seems to slow down and you instinctively dodge!" ) );
-        } else if( seen ) {
-            add_msg( _( "%s dodges… so fast!" ), who.disp_name() );
-        }
-        return true;
-    } else {
-        if( is_u ) {
-            add_msg( _( "You try to dodge but there's no room!" ) );
-        } else if( seen ) {
-            add_msg( _( "%s tries to dodge but there's no room!" ), who.disp_name() );
+
+    const auto result = uncanny_dodge_result( who );
+    const bool is_u = who.is_avatar();
+    const bool seen = is_u || get_player_character().sees( who );
+
+    who.mod_power_level( -bio_uncanny_dodge->power_trigger );
+
+    const auto visitor = cata::match{
+        [&who, is_u, seen]( const tripoint & dest )
+        {
+            who.mod_stored_kcal( -trigger_rcs_cost_kcal );
+            if( is_u ) {
+                add_msg( _( "Time seems to slow down and you're ejected to safe space!" ) );
+            } else if( seen ) {
+                add_msg( _( "%s dodges so fast as if there's rocket attached behind!" ), who.disp_name() );
+            }
+            auto &here = get_map();
+            here.add_field( who.pos(), fd_smoke, 1 );
+            who.setpos( dest );
+            return true;
+        },
+        [&who, is_u, seen]( const UncannyDodgeStatus fail ) -> bool {
```

Comment:
...I think this should eliminatethe rcs check correctly but please double-check my work on this function in particular beforehand should you opt to implement this.
```suggestion
    const auto visitor = cata::match{
        [&who, is_u, seen]( const UncannyDodgeStatus fail ) -> bool {
```